internal/rpc/eth-evm.go

```diff
@@ -0,0 +1,130 @@
+package rpc
+
+import (
+	"context"
+	"log/slog"
+	"sync"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+var (
+	ethEvmRpcs = make(map[string]EthEvmRpc)
+)
+
+const retriesKey = contextKey("retries")
+
+type contextKey string
+
+// EthEvmRpc is the interface for the Ethereum RPC client.
+type EthEvmRpc interface {
+	BlockNumber(ctx context.Context) (uint64, error)
+	FilterLogs(ctx context.Context, query ethereum.FilterQuery) ([]types.Log, error)
+	TransactionByHash(ctx context.Context, hash common.Hash) (*types.Transaction, bool, error)
+}
+
+type ethEvmRpc struct {
+	network string
+	mutex   sync.Mutex
+	client  *ethclient.Client
+	retries int
+}
+
+func NewEthEvmRpc(network string, url string, retries int) (EthEvmRpc, error) {
+	if _, ok := ethEvmRpcs[network]; ok {
+		return ethEvmRpcs[network], nil
+	}
+	slog.Debug("initializing new eth-evm rpc |", "network", network)
+	client, err := ethclient.Dial(url)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Unfortunately, `Dial` only checks for a successful network protocol connection. It does not check if the endpoint is indeed an RPC node on the correct chain. A check to validate this stronger can be found [here](https://github.com/NethermindEth/sedge/blob/feat/lido-exporter/internal/lido/contracts/client.go#L35-L51).

Reply:
If the URL isn’t an RPC, the exporter will fail, and the `eoe_eigenda_exporter_up` will equal 0. We should be able to receive an alert for it and check the logs to see if there is an RPC error.

Reply:
An RPC call would fail but you catch the problem later and you might get a 4XX error or something else. Even worse if you use another chain RPC. Validating the RPC as soon as possible lets you report a very specific error, enhancing the debug and avoiding potential side effects.

If eoe fails after this check, the up metric will be zero anyway.

Reply:
ChainID for Mainnet should be `1`. You might already have this change addressed but I'm highlighting this here to track it.

Reply:
FIxed https://github.com/NethermindEth/eigenlayer-onchain-exporter/pull/3/commits/4918f454eb1fefdaf4f77c6727c277be7613ada4